src/com/google/firebase/FirestoreDAO.js

```diff
@@ -0,0 +1,221 @@
+          config.keyFilename = this.gcloudCredentialsPath;
+        return new this.firestoreModule(config);
+      },
+    },
+    {
+      name: 'collection',
+      documentation: `The firebase.firestore.CollectionReference for this DAO:
+          https://firebase.google.com/docs/reference/js/firebase.firestore.CollectionReference`,
+      transient: true,
+      factory: function() {
+        return this.firestore.collection(this.collectionPath);
+      },
+    },
+    {
+      // class: 'FObjectProperty',
+      // of: 'foam.mlang.F',
+      name: 'firestoreDocumentID',
+      documentation: `The function used to extract a Firestore Document ID string
+          from a foam.core.FObject that is stored via this DAO.`,
+      factory: function() { return this.DefaultFirestoreDocumentID.create(); }
+    },
+    {
+      // class: 'FObjectProperty',
+      // of: 'foam.mlang.F',
+      name: 'firestoreData',
+      documentation: `The function used to extract a Firestore data from a
+          foam.core.FObject that is stored via this DAO.`,
+      factory: function() { return this.DefaultFirestoreData.create(); }
+    },
+    {
+      // class: 'FObjectProperty',
+      // of: 'foam.mlang.F',
+      name: 'fobject',
+      documentation: `The function used to extract a foam.core.FObject from
+          Firestore data that is stored via this DAO.`,
+      factory: function() { return this.DefaultFObject.create(); }
+    },
+    {
+      class: 'Array',
+      name: 'putBacklog_',
+    }
+  ],
+
+  methods: [
+    function put_(x, obj) {
+      var resolve;
+      var reject;
+      var promise = new Promise(function(res, rej) {
+        resolve = res;
+        reject = rej;
+      });
+      this.putBacklog_.push({
+        obj: obj,
+        resolve: resolve,
+        reject: reject,
+        promise: promise,
+      });
+      if ( this.putBacklog_.length > 400 ) {
+        this.putBatch_();
+      }
+      this.onPut();
+      return promise;
+    },
+    function putBatch_() {
+      var batch = this.firestore.batch();
+      var backlog = this.putBacklog_;
+      for ( var i = 0; i < backlog.length; i++ ) {
+        var data = backlog[i];
+        batch.set(this.getDoc_(data.obj), this.firestoreData.f(data.obj));
+      }
+
+      batch.commit().then(function(backlog) {
+        for ( var i = 0; i < backlog.length; i++ ) {
+          backlog[i].resolve(backlog[i].obj);
+        }
+      }.bind(this, backlog), function(error) {
+        for ( var i = 0; i < backlog.length; i++) {
+          backlog[i].reject(error);
+        }
+      });
+
+      this.putBacklog_ = [];
+    },
+    function remove_(x, obj) {
+      return this.getDoc_(obj).delete();
+    },
+    function find_(x, idOrObj) {
+      return (foam.core.FObject.isInstance(idOrObj) ?
+              this.getDoc_(idOrObj) : this.collection.doc(idOrObj.toString()))
+          .get()
+          .then(function(docSnapshot) { return docSnapshot.data() || null; });
```

Comment:
Not running this through `this.fobject.f`?

Reply:
That's a bug. Good catch. Fix pushed.